add remediation hint to ha cluster check

idevat opened the pull request:
This pull request is based on https://projects.engineering.redhat.com/browse/OAMG-3591

examon commented:
leapp-ci build

ZhukovGreen commented on repos/system_upgrade/el7toel8/actors/checkhacluster/libraries/checkhacluster.py:
```diff
@@ -25,6 +25,17 @@ def inhibit(node_type):
         ),
+        reporting.Remediation(
+            hint=(
+                "Destroy the existing HA cluster"
```
Hi @idevat , I might be wrong, but is there a way to check if the cluster is disabled from inside the actor? Then we could just update the condition to inhibit the upgrade (add the check of cluster state), rather than asking the user to remove the config files

idevat replied:
Red Hat does not support in-place upgrades or rolling-upgrades of cluster nodes from one major release of RHEL to another. Upgrade cannot continue if HA cluster exists.

ZhukovGreen replied:
OK, then, is there a better way to check if HA cluster enabled (i.e. pcs cluster status) instead of a check of the existence of configs?

idevat replied:
Why you want only to disable HA cluster? The configuration from RHEL 7 will not properly work in RHEL 8. The HA cluster must be destroyed.

ZhukovGreen replied:
Synced over chat. As I understood we can't do better and need to check configs files and if they exist, we have to ask the user to remove them as well.

ZhukovGreen reviewed (commented):
Hi @idevat than you for your contribution,

just a little question to your MR.

ZhukovGreen reviewed (commented):
Just a short question...

ZhukovGreen commented:
@oamg/qe guys, do we need to verify this?

bocekm commented:
Yes, the verification is going to happen - tracking it under OAMG-3591.

ZhukovGreen commented:
leapp-ci build

ZhukovGreen commented:
e2e tests

idevat commented:
Is there anything else I should do? I see that leapp-ci tests failed, however it seems that `checkhacluster` passed and failed another actors.

bocekm commented:
@AloisMahdal, @mkluson: please try to make sure the upgrade test for this is fine.

vojtechsokol commented:
@idevat Hi, there were some updates to our linter process so please rebase the PR and the fails for other actors should be gone.

AloisMahdal commented:
e2e tests

idevat commented:
It seems we are waiting for something. Is there anything I should do?

bocekm commented:
@idevat, we wait for @AloisMahdal to finish the verification. See https://issues.redhat.com/browse/OAMG-3827 for details.

idevat commented:
@bocekm, no problem. I just wanted to make sure that it is not waiting for me.

AloisMahdal commented:
Verifying on x86_64.

Current master build:

    Risk Factor: high (inhibitor)
    Title: Use of HA cluster detected. Upgrade can't proceed.
    Summary: HA cluster is not supported by the inplace upgrade.
    HA cluster configuration file(s) found. It seems to be a cluster node.

This PR, ie. *leapp-repository-0.11.0-0.202010050755Z.972f128.972f128f_master.PR578.el7_6.noarch*:

    Risk Factor: high (inhibitor)
    Title: Use of HA cluster detected. Upgrade can't proceed.
    Summary: HA cluster is not supported by the inplace upgrade.
    HA cluster configuration file(s) found. It seems to be a cluster node.
    Remediation: [hint] Destroy the existing HA cluster or (if you have already removed HA cluster packages) remove configuration files /var/lib/pacemaker/cib/cib.xml and /etc/corosync/corosync.conf
    [command] pcs cluster destroy --all

after calling

    pcs cluster stop --all --wait
    pcs cluster destroy --all

the inhibitor is not re-raised.

The question is, shouldn't the remediation contain both commands?

(If yes, note that because of current leapp policy it has to be in single item of the `commands` list, ie. joined by eg. `&&`.)

AloisMahdal commented:
(Note that I have those commands from clusterha QE; I have no idea if both are really needed.)

idevat commented:
The command `pcs cluster destroy --all` stops the cluster first. However, it is done by slightly different way than `pcs cluster stop  --all --wait` and it could be safer in some corner cases to use `pcs cluster stop...` first. I've added `pcs cluster stop...` to a remediation command and also rebased it on the current master branch.

AloisMahdal commented:
As of *leapp-repository-0.11.0-0.202010141246Z.327580b.PR_master.PR578.el7_6.noarch*, the text is:

    Risk Factor: high (inhibitor)
    Title: Use of HA cluster detected. Upgrade can't proceed.
    Summary: HA cluster is not supported by the inplace upgrade.
    HA cluster configuration file(s) found. It seems to be a cluster node.
    Remediation: [hint] Destroy the existing HA cluster or (if you have already removed HA cluster packages) remove configuration files /var/lib/pacemaker/cib/cib.xml and /etc/corosync/corosync.conf
    [command] pcs cluster stop --all --wait && pcs cluster destroy --all
    ----------------------------------------

and the behavior is as expected.

IOW, this is verified.

pirat89 commented:
@idevat @AloisMahdal does this works well using cockpit or satellite to execute the command? Regarding change proposed by @shaded-enmity  it seems that it could be still understood as "two commands" by satellite/cockpit. not sure. see https://github.com/oamg/leapp-repository/pull/540 - focus on change where commands are packed to be executed like: `sh -c "cmd && cmd ..."

AloisMahdal commented:
Not sure; JSON of the above:

    [
      {
        "type": "hint",
        "context": "Destroy the existing HA cluster or (if you have already removed HA cluster packages) remove configuration files /var/lib/pacemaker/cib/cib.xml and /etc/corosync/corosync.conf"
      },
      {
        "type": "command",
        "context": [
          "pcs",
          "cluster",
          "stop",
          "--all",
          "--wait",
          "&&",
          "pcs",
          "cluster",
          "destroy",
          "--all"
        ]
      }
    ]

shaded-enmity commented:
@pirat89 is right, remediation are not executed via shell, so if we need to run 2 commands it should be `sh -c "command1 && command2"`

pirat89 commented on repos/system_upgrade/el7toel8/actors/checkhacluster/libraries/checkhacluster.py:
```diff
@@ -25,6 +25,17 @@ def inhibit(node_type):
+                    COROSYNC_CONF_LOCATION,
+                )
+            ),
+            commands=[["pcs", "cluster", "stop", "--all", "--wait", "&&", "pcs", "cluster", "destroy", "--all"]]
```
@idevat  It there is not typo, this should work:
```suggestion
            commands=[["sh", "-c", "pcs cluster stop --all --wait && pcs cluster destroy --all"]]
```

idevat replied:
I'll try it...

pirat89 reviewed (approved):
Thanks!

AloisMahdal commented:
So, with *leapp-repository-0.11.0-0.202010191256Z.6e771e9.PR_master.PR578.el7_6.noarch*..

leapp-report.txt:

    Risk Factor: high (inhibitor)
    Title: Use of HA cluster detected. Upgrade can't proceed.
    Summary: HA cluster is not supported by the inplace upgrade.
    HA cluster configuration file(s) found. It seems to be a cluster node.
    Remediation: [hint] Destroy the existing HA cluster or (if you have already removed HA cluster packages) remove configuration files /var/lib/pacemaker/cib/cib.xml and /etc/corosync/corosync.conf
    [command] sh -c pcs cluster stop --all --wait && pcs cluster destroy --all
    ----------------------------------------

leapp-report.json:

    [
      {
        "type": "hint",
        "context": "Destroy the existing HA cluster or (if you have already removed HA cluster packages) remove configuration files /var/lib/pacemaker/cib/cib.xml and /etc/corosync/corosync.conf"
      },
      {
        "type": "command",
        "context": [
          "sh",
          "-c",
          "pcs cluster stop --all --wait && pcs cluster destroy --all"
        ]
      }
    ]

But the remediation does not really work :/  when I click remediate in cockpit I get pcs usage message:ZZ

![ha-remediate](https://user-images.githubusercontent.com/217935/96583459-0839a500-12dd-11eb-8064-10e95431113a.png)

idevat commented:
I see it. The command is `sh -c pcs cluster stop --all --wait && pcs cluster destroy --all` but is should be `sh -c "pcs cluster stop --all --wait && pcs cluster destroy --all"`. I will try fixing it...

AloisMahdal commented:
According to @shaded-enmity the issue shown on the screenshot is entirely issue of cockpit so we can ignore it for purposes of this PR.

shaded-enmity commented:
The current data is alright, this will be fixed in Cockpit.

pirat89 commented:
Based on discussion above and on the mtg, merging.
